### PR TITLE
refactor: use workspaces to run test and fix hoisted deps

### PR DESCRIPTION
**Pull request by fengmk2**
##### Checklist

- [ ] `npm test` passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows commit guidelines

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -24,5 +24,8 @@
     "resolveJsonModule": true,
     "emitDecoratorMetadata": true,
     "experimentalDecorators": true
+  },
+  "ts-node": {
+    "files": true
```
Comment (fengmk2) [inline tsconfig.json:29]: https://typestrong.org/ts-node/docs/options#files 好难。。。

Comment (fengmk2) [reply tsconfig.json:29]: egg-bin 默认内置 https://github.com/eggjs/egg-bin/commit/35e89dbdbfcb6d2c6cd07f73145ead7c4c5421ce#diff-d8db36db6bfb32591336006f1dc1adee9e2caf7b021469995aa687c5c5723d0cR59

**Comment by fengmk2** [comment]
@killagu 终于搞完，之前省略了类型检查。。。

发布还是走了 lerna，然后 test 走了 npm workspaces

File: core/runtime/src/model/ContextHandler.ts [hunk 2]
```diff
@@ -8,8 +8,7 @@ export class ContextHandler {
   static runInContextCallback: runInContextCallback;
 
   static getContext(): EggContext | undefined {
-    assert(this.getContextCallback, 'getContextCallback not set');
```
Comment (fengmk2) [inline core/runtime/src/model/ContextHandler.ts:11]: standalone/standalone 的测试用例跑的时候，getContextCallback 会是 undefined 的。

File: plugin/controller/package.json [hunk 3]
```diff
@@ -67,6 +67,7 @@
     "egg": "^3.9.1",
     "egg-mock": "^5.5.0",
     "egg-tracer": "^1.1.0",
+    "globby": "^11.1.0",
```
Comment (fengmk2) [inline plugin/controller/package.json:70]: 缺少依赖

File: core/metadata/package.json [hunk 4]
```diff
@@ -41,6 +41,7 @@
     "egg-errors": "^2.2.3"
   },
   "devDependencies": {
+    "globby": "^11.1.0",
```
Comment (fengmk2) [inline core/metadata/package.json:44]: 缺少依赖

File: core/eventbus-runtime/package.json [hunk 5]
```diff
@@ -41,8 +40,7 @@
     "@eggjs/core-decorator": "^3.1.0",
     "@eggjs/eventbus-decorator": "^3.1.0",
     "await-event": "^2.1.0",
-    "await-first": "^1.0.0",
-    "typed-emitter": "^1.3.1"
```
Comment (fengmk2) [inline core/eventbus-runtime/package.json:45]: 多余依赖 typed-emitter

File: core/aop-runtime/package.json [hunk 6]
```diff
@@ -53,6 +53,7 @@
   },
   "devDependencies": {
     "@eggjs/module-test-util": "^3.2.1",
-    "@eggjs/tegg-loader": "^3.1.0"
+    "@eggjs/tegg-loader": "^3.1.0",
+    "mm": "^3.2.1"
```
Comment (fengmk2) [inline core/aop-runtime/package.json:57]: 缺少依赖

File: package.json [hunk 7]
```diff
@@ -17,36 +17,36 @@
   "scripts": {
     "lint:fix": "eslint . --ext .ts --fix",
     "lint": "eslint . --ext .ts",
-    "test": "cross-env NODE_ENV=test NODE_OPTIONS='--no-deprecation' TS_NODE_TRANSPILE_ONLY=true mocha",
```
Comment (fengmk2) [inline package.json:20]: TS_NODE_TRANSPILE_ONLY=true 不能随便设置。。。

File: core/runtime/src/model/ContextHandler.ts (hunk 2 above)
Comment (killagu) [reply core/runtime/src/model/ContextHandler.ts:11]: 这个是 bug 了，我来修复下。

File: core/eventbus-runtime/package.json [hunk 8]
```diff
@@ -40,18 +39,17 @@
   "dependencies": {
     "@eggjs/core-decorator": "^3.1.0",
     "@eggjs/eventbus-decorator": "^3.1.0",
+    "@eggjs/tegg-common-util": "^1.2.3",
+    "@eggjs/tegg-runtime": "^3.2.1",
```
Comment (fengmk2) [inline core/eventbus-runtime/package.json:43]: 错误加到 devDependencies 了

File: plugin/controller/package.json [hunk 9]
```diff
@@ -47,6 +47,7 @@
   },
   "dependencies": {
     "@eggjs/egg-module-common": "^1.0.0",
+    "@eggjs/router": "^2.0.1",
```
Comment (fengmk2) [inline plugin/controller/package.json:50]: 错误加到 devDependencies 了

**Comment by killagu** [comment]
- @eggjs/aop-decorator@3.2.2
 - @eggjs/tegg-aop-runtime@3.2.2
 - @eggjs/tegg-background-task@3.2.2
 - @eggjs/tegg-common-util@3.2.2
 - @eggjs/controller-decorator@3.2.2
 - @eggjs/core-decorator@3.2.2
 - @eggjs/tegg-dynamic-inject-runtime@3.2.2
 - @eggjs/tegg-dynamic-inject@3.2.2
 - @eggjs/eventbus-decorator@3.2.2
 - @eggjs/tegg-eventbus-runtime@3.2.2
 - @eggjs/tegg-lifecycle@3.2.2
 - @eggjs/tegg-loader@3.2.2
 - @eggjs/tegg-metadata@3.2.2
 - @eggjs/tegg-orm-decorator@3.2.2
 - @eggjs/tegg-runtime@3.2.2
 - @eggjs/tegg-schedule-decorator@3.2.2
 - @eggjs/standalone-decorator@3.2.2
 - @eggjs/tegg@3.2.2
 - @eggjs/tegg-aop-plugin@3.2.2
 - @eggjs/egg-module-common@3.2.2
 - @eggjs/tegg-config@3.2.2
 - @eggjs/tegg-controller-plugin@3.2.2
 - @eggjs/tegg-eventbus-plugin@3.2.2
 - @eggjs/tegg-orm-plugin@3.2.2
 - @eggjs/tegg-schedule-plugin@3.2.2
 - @eggjs/tegg-plugin@3.2.2
 - @eggjs/tegg-standalone@3.2.2

**Pull request merged**